File: android/app/src/main/java/com/clock3/pet/data/model/Alarm.kt
package com.clock3.pet.data.model

import com.clock3.pet.data.entity.AlarmEntity
import java.time.DayOfWeek
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.format.DateTimeFormatter

data class Alarm(
    val id: Long = 0,
    val label: String,
    val time: String,
    val content: String = "",
    val repeatType: RepeatType = RepeatType.ONCE,
    val repeatDays: List<Int> = emptyList(),
    val enabled: Boolean = true,
    val snoozeCount: Int = 0,
    val snoozeTime: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    enum class RepeatType(val value: String) {
        ONCE("once"),
        DAILY("daily"),
        WORKDAYS("workdays"),
        WEEKEND("weekend"),
        CUSTOM("custom");

        companion object {
            fun fromValue(value: String): RepeatType {
                return entries.find { it.value == value } ?: ONCE
            }
        }
    }

    fun getNextTriggerTime(): LocalDateTime? {
        val now = LocalDateTime.now()
        val parts = time.split(":")
        if (parts.size != 2) return null

        val hour = parts[0].toIntOrNull()
        val minute = parts[1].toIntOrNull()
        if (hour == null || minute == null || hour !in 0..23 || minute !in 0..59) return null
        val targetTime = LocalTime.of(hour, minute)
        var target = LocalDateTime.of(now.toLocalDate(), targetTime)

        if (target.isBefore(now) || target.isEqual(now)) {
            target = target.plusDays(1)
        }

        val maxAttempts = 365

        when (repeatType) {
            RepeatType.ONCE -> {
                return if (target.isAfter(now)) target else null
            }
            RepeatType.DAILY -> {
                return target
            }
            RepeatType.WORKDAYS -> {
                var attempts = 0
                while (attempts < maxAttempts) {
                    if (target.dayOfWeek in WEEKDAYS && target.isAfter(now)) {
                        return target
                    }
                    target = target.plusDays(1)
                    attempts++
                }
                return null
            }
            RepeatType.WEEKEND -> {
                var attempts = 0
                while (attempts < maxAttempts) {
                    if (target.dayOfWeek in WEEKEND_DAYS && target.isAfter(now)) {
                        return target
                    }
                    target = target.plusDays(1)
                    attempts++
                }
                return null
            }
            RepeatType.CUSTOM -> {
                if (repeatDays.isEmpty()) return null
                var attempts = 0
                while (attempts < maxAttempts) {
                    if (repeatDays.contains(target.dayOfWeek.value) && target.isAfter(now)) {
                        return target
                    }
                    target = target.plusDays(1)
                    attempts++
                }
                return null
            }
        }
    }

    fun toEntity(): AlarmEntity {
        return AlarmEntity(
            id = id,
            label = label,
            time = time,
            content = content,
            repeatType = repeatType.value,
            repeatDays = if (repeatDays.isNotEmpty()) repeatDays.joinToString(",") else null,
            enabled = enabled,
            snoozeCount = snoozeCount,
            snoozeTime = snoozeTime,
            createdAt = createdAt,
            updatedAt = updatedAt
        )
    }

    fun toExportMap(): Map<String, Any> {
        return mapOf(
            "id" to id,
            "label" to label,
            "time" to time,
            "content" to content,
            "repeatType" to repeatType.value,
            "repeatDays" to repeatDays,
            "enabled" to enabled,
            "snoozeCount" to snoozeCount,
            "snoozeTime" to (snoozeTime ?: ""),
            "createdAt" to (createdAt ?: ""),
            "updatedAt" to (updatedAt ?: "")
        )
    }

    companion object {
        private val WEEKDAYS = listOf(DayOfWeek.MONDAY, DayOfWeek.TUESDAY, DayOfWeek.WEDNESDAY,
            DayOfWeek.THURSDAY, DayOfWeek.FRIDAY)
        private val WEEKEND_DAYS = listOf(DayOfWeek.SATURDAY, DayOfWeek.SUNDAY)

        fun fromEntity(entity: AlarmEntity): Alarm {
            return Alarm(
                id = entity.id,
                label = entity.label,
                time = entity.time,
                content = entity.content,
                repeatType = RepeatType.fromValue(entity.repeatType),
                repeatDays = entity.repeatDays?.split(",")?.mapNotNull { it.toIntOrNull() } ?: emptyList(),
                enabled = entity.enabled,
                snoozeCount = entity.snoozeCount,
                snoozeTime = entity.snoozeTime,
                createdAt = entity.createdAt,
                updatedAt = entity.updatedAt
            )
        }

        fun fromExportMap(map: Map<*, *>): Alarm {
            return Alarm(
                id = 0,
                label = (map["label"] as? String) ?: "闹钟",
                time = (map["time"] as? String) ?: "08:00",
                content = (map["content"] as? String) ?: "",
                repeatType = (map["repeatType"] as? String)?.let { RepeatType.fromValue(it) } ?: RepeatType.ONCE,
                repeatDays = (map["repeatDays"] as? List<*>)?.mapNotNull { (it as? Number)?.toInt() } ?: emptyList(),
                enabled = (map["enabled"] as? Boolean) ?: true,
                snoozeCount = (map["snoozeCount"] as? Number)?.toInt() ?: 0,
                snoozeTime = (map["snoozeTime"] as? String)?.let { if (it.isNotEmpty()) it else null },
                createdAt = (map["createdAt"] as? String)?.let { if (it.isNotEmpty()) it else null },
                updatedAt = (map["updatedAt"] as? String)?.let { if (it.isNotEmpty()) it else null }
            )
        }
    }
}

File: app/models/alarm.py
from datetime import datetime
from typing import Optional, List
import json

class Alarm:
    """闹钟数据模型"""
    
    def __init__(
        self,
        id: Optional[int] = None,
        label: str = "",
        time: str = "08:00",
        content: str = "",
        repeat_type: str = "once",
        repeat_days: Optional[List[int]] = None,
        enabled: bool = True,
        snooze_count: int = 0,
        snooze_time: Optional[str] = None,
        created_at: Optional[str] = None,
        updated_at: Optional[str] = None
    ):
        self.id = id
        self.label = label
        self.time = time
        self.content = content
        self.repeat_type = repeat_type
        self.repeat_days = repeat_days or []
        self.enabled = enabled
        self.snooze_count = snooze_count
        self.snooze_time = snooze_time
        self.created_at = created_at
        self.updated_at = updated_at
    
    def get_next_trigger_time(self) -> Optional[datetime]:
        """获取下次触发时间"""
        now = datetime.now()
        hour, minute = map(int, self.time.split(':'))
        
        target = now.replace(hour=hour, minute=minute, second=0, microsecond=0)
        
        if self.repeat_type == 'once':
            if target <= now:
                return None
            return target
        
        elif self.repeat_type == 'daily':
            if target <= now:
                target = target + timedelta(days=1)
            return target
        
        elif self.repeat_type == 'workdays':
            weekdays = [1, 2, 3, 4, 5]
            while True:
                if target.weekday() in weekdays and target > now:
                    return target
                target = target + timedelta(days=1)
        
        elif self.repeat_type == 'weekend':
            while True:
                if target.weekday() in [5, 6] and target > now:
                    return target
                target = target + timedelta(days=1)
        
        elif self.repeat_type == 'custom' and self.repeat_days:
            while True:
                if target.weekday() in self.repeat_days and target > now:
                    return target
                target = target + timedelta(days=1)
        
        return None
    
    def should_trigger_today(self) -> bool:
        """检查今天是否应该触发"""
        if self.repeat_type == 'once':
            return True
        elif self.repeat_type == 'daily':
            return True
        elif self.repeat_type == 'workdays':
            return datetime.now().weekday() in [1, 2, 3, 4, 5]
        elif self.repeat_type == 'weekend':
            return datetime.now().weekday() in [5, 6]
        elif self.repeat_type == 'custom':
            return datetime.now().weekday() in (self.repeat_days or [])
        return False
    
    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            'id': self.id,
            'label': self.label,
            'time': self.time,
            'content': self.content,
            'repeat_type': self.repeat_type,
            'repeat_days': self.repeat_days,
            'enabled': self.enabled,
            'snooze_count': self.snooze_count,
            'snooze_time': self.snooze_time,
            'created_at': self.created_at,
            'updated_at': self.updated_at
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Alarm':
        """从字典创建"""
        return cls(
            id=data.get('id'),
            label=data.get('label', ''),
            time=data.get('time', '08:00'),
            content=data.get('content', ''),
            repeat_type=data.get('repeat_type', 'once'),
            repeat_days=data.get('repeat_days'),
            enabled=data.get('enabled', True),
            snooze_count=data.get('snooze_count', 0),
            snooze_time=data.get('snooze_time'),
            created_at=data.get('created_at'),
            updated_at=data.get('updated_at')
        )
    
    def __repr__(self):
        return f"Alarm(id={self.id}, label={self.label}, time={self.time}, enabled={self.enabled})"
